tree/treeplayer/src/TTreeReaderValue.cxx

```diff
@@ -133,9 +133,67 @@ void ROOT::Internal::TTreeReaderValueBase::RegisterWithTreeReader() {
+{
+   if ((fProxy->*Func)()) {
+      fReadStatus = kReadSuccess;
+   } else {
+      fReadStatus = kReadError;
+   }
+   return fReadStatus;
+}
+
 ROOT::Internal::TTreeReaderValueBase::EReadStatus
-ROOT::Internal::TTreeReaderValueBase::ProxyRead() {
+ROOT::Internal::TTreeReaderValueBase::ProxyReadDefaultImpl() {
    if (!fProxy) return kReadNothingYet;
+   if (fProxy->IsInitialized() || fProxy->Setup()) {
+
+      using EReadType = ROOT::Detail::TBranchProxy::EReadType;
+      using TBranchPoxy = ROOT::Detail::TBranchProxy;
+
+      EReadType readtype = EReadType::kNoDirector;
+      if (fProxy) readtype = fProxy->GetReadType();
+
+      switch (readtype) {
+         case EReadType::kNoDirector:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadNoDirector>;
+            break;
+         case EReadType::kReadParentNoCollection:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadParentNoCollection>;
+            break;
+         case EReadType::kReadParentCollectionNoPointer:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadParentCollectionNoPointer>;
+            break;
+         case EReadType::kReadParentCollectionPointer:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadParentCollectionPointer>;
+            break;
+         case EReadType::kReadNoParentNoBranchCountCollectionPointer:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadNoParentNoBranchCountCollectionPointer>;
+            break;
+         case EReadType::kReadNoParentNoBranchCountCollectionNoPointer:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadNoParentNoBranchCountCollectionNoPointer>;
+            break;
+         case EReadType::kReadNoParentNoBranchCountNoCollection:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadNoParentNoBranchCountNoCollection>;
+            break;
+         case EReadType::kReadNoParentBranchCountCollectionPointer:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadNoParentBranchCountCollectionPointer>;
+            break;
+         case EReadType::kReadNoParentBranchCountCollectionNoPointer:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadNoParentBranchCountCollectionNoPointer>;
+            break;
+         case EReadType::kReadNoParentBranchCountNoCollection:
+            fProxyReadFunc = &TTreeReaderValueBase::ProxyReadTemplate<&TBranchPoxy::ReadNoParentBranchCountNoCollection>;
+            break;
+         case EReadType::kDefault:
+            // intentional fall through.
+         default: fProxyReadFunc = &TTreeReaderValueBase::ProxyReadDefaultImpl;
+      }
+      return (this->*fProxyReadFunc)();
+   }
```

Comment:
Could you add a comment in which cases this `if` block is not executed, i.e. in which cases `if (fProxy->Read()) {` below is hit? The code reads as if this if is guarding a failure to set up the proxy, and if it fails, we call `if (fProxy->Read()) {`.

Reply:
humm ... it actually looks like I forgot to remove that 'now-unrecheable' code anyway ....

Reply:
it is there to avoid duplicating the error handling (error printing).